include/boost/gil/image_processing/morph.hpp

```diff
@@ -0,0 +1,106 @@
+//
+// Copyright 2019 Ayush Bansal <abansal1008@gmail.com>
+//
+// Use, modification and distribution are subject to the Boost Software License,
+// Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+#ifndef BOOST_GIL_IMAGE_PROCESSING_MORPH_HPP
+#define BOOST_GIL_IMAGE_PROCESSING_MORPH_HPP
+
+#include <boost/assert.hpp>
+#include <boost/gil/image_view_factory.hpp>
+#include <boost/gil.hpp>
+#include <boost/multi_array.hpp>
```

Comment:
I actually wanted to discuss this. I wanted to know your opinion on what data structure should we use to define a Structuring Element. Any binary matrix like structure would work just fine. Do you think using a simple array/vector here would be more appropriate?

Reply:
I agree with @mloskot 's objection, and I admit I'm a bit confused after a (rather quick, I admit) look at this PR: you include the `boost/multi_array.hpp` header, but you never use it (explicitly).
Rather, you use a template argument `SE`, which in itself sounds perfectly fine, but I think requires to specify exactly what kind of requirements you put onto that type.
Could we get away with using the same requirements we put on `SrcView` ?

Reply:
Ohh, yes it was a mistake to have included `boost/multi_array.hpp` in current code. I was actually playing around with multi_array to figure out a way to put up a requirement check on it and I forgot to remove that header in the end. However, in the next iteration I would be requiring it as i have included this check - 
`boost::function_requires<boost::multi_array_concepts::ConstMultiArrayConcept<StructElement, 2>>();`
@mloskot  @stefanseefeld could you please give your comments on whether using multi_array as the type of SE and using this check is sufficient and correct?

Reply:
Well, as I said, I think requiring multi_array would be wrong. The whole point of parametrizing these functions with `SE` is to let users fill in that parameter with their own type, so long as it meets the requirements.
(For avoidance of doubt: using multi_array in the example is perfectly fine.)

Reply:
@ayushbansal07
> could you please give your comments on whether using multi_array as the type of SE and using this check is sufficient and correct?

As I poined earlier, I also agree with @stefanseefeld on this: please try to avoid `multi_array` in codebase of GIL core and tests, and try to make those possibly container-agnostic. In examples, it is fine.

Reply:
@mloskot @stefanseefeld I understand your point now. I have updated morphology.hpp. I now model a SE to be of type vector or any similar structure (which can support .size() functionality).

Reply:
@ayushbansal07 
> I now model a SE to be of type vector or any similar structure (which can support .size() functionality).

As per one of the assertions, there is also required subscript (index) operator on element

Reply:
@mloskot 
I am unable to find any check for subscript operator. I however did find a check for Container type.I wondered if checking for a Container type. So would using that and replacing all my occurrences of `elem[i]` with `elem.begin()+i` be a good thing to do?

Reply:
Ideally if there is a concept representing requirements of `StructElement`, but I wouldn't bother for now.

For now, I'd simply add (Doxygen-style) comments to `erode` and `dilate` explaining the requirement, similar to https://github.com/boostorg/gil/blob/8235ee645a4e8c6e20722cc9c9eebe1f8050f9b6/include/boost/gil/extension/numeric/kernel.hpp#L32

In future, structuring element could be defined similarly to the kernel classes.
Since your algorithm now sort of assumes rectangular shape of the structured element, it could reuse the `kernel_2d`, but I don't think it is necessary or even a good idea considering future improvements (e.g. ovals and other non-rectangular shape of structuring elements 😄).

Reply:
FYI, I'm marking this conversation as resolved